File: examples/rmi/Printer/cpp/src/PrinterImpl.cpp
/*
 * PrinterImpl.cpp
 *
 *  Created on: 7 juin 2012
 *      Author: nawel
 */

#include "PrinterImpl.h"
#include "vortex_os.h"


int PrinterImpl::id;

PrinterImpl::PrinterImpl(std::string name, std::string location, bool color)
{
  infos.name = name.c_str();
  infos.location = location.c_str();
  infos.s = P_IDLE;
  infos.color = color;
  activeJobId = -1;
}

PrinterImpl::~PrinterImpl()
{
}

printer::PrinterInfos
PrinterImpl::getPrinterInfos()
{
  return infos; // may be explicit copy is needed ...
}

printer::JobReport
PrinterImpl::printDoc(DDS::String doc_name, DDS::Boolean color,
    DDS::Long & jobId)
{
  JobReport report;
  report.printer_name = DDS::string_dup(infos.name);
  report.doc_name = DDS::string_dup(doc_name);
  report.nbPrintedPages = 0;
  report.printer_result = PRINTER_OK;
  report.message = DDS::string_dup("Succeeded in printing the document !");
  jobId = 0;

  if (infos.s != P_IDLE)
    {
      report.printer_result = UNAVAILABLE;
      report.message = DDS::string_dup("The printer is still busy !");
      return report;
    }
  else if (color && !infos.color)
    {
      report.printer_result = NO_COLOR;
      report.message = DDS::string_dup("The printer is black & white !");
      return report;
    }

  if (strlen(doc_name) > 20)
    {
      report.printer_result = TOO_BIG;
      report.message = DDS::string_dup("The document is too big !");
      return report;
    }

  setPrinterStatus(P_BUSY);
  jobId = getJobId();
  if (jobId % 7 == 0)
    {
      setPrinterStatus(P_IDLE);
      report.printer_result = ERR;
      report.message = DDS::string_dup("Printer Error occured !");
    }

  std::cout << "Printing document " << doc_name << " ... " << std::endl;
  setActiveJob(jobId);
  os_time delay =
    { jobId, 0 };
  os_nanoSleep(delay);
  setPrinterStatus(P_IDLE);
  report.nbPrintedPages = strlen(doc_name);
  setActiveJob(-1);
  return report;
}

void
PrinterImpl::setPrinterStatus(Status s)
{
  Locker lock(mutex);
  infos.s = s;
}

Status
PrinterImpl::getPrinterStatus()
{
  Locker lock(mutex);
  return infos.s;
}

void
PrinterImpl::resume()
{
  std::cout << "Printer " << infos.name << " activated." << std::endl;
}

void
PrinterImpl::suspend()
{
  os_time delay_2s = { 2, 0 };
  while (getPrinterStatus() != P_IDLE)
    {
      std::cout << "Waiting for the current job to complete ..." << std::endl;
      os_nanoSleep(delay_2s);
    }
  std::cout << "Printer " << infos.name << " stopped." << std::endl;
}

DDS::Long
PrinterImpl::getActiveJob()
{
  Locker lock(mutex);
  return activeJobId;
}

void
PrinterImpl::setActiveJob(DDS::Long id)
{
  Locker lock(mutex);
  activeJobId = id;
}

int
PrinterImpl::getJobId()
{
  return PrinterImpl::id++;
}

File: examples/rmi/Printer/java/src/dds/rmi/example/printer/PrinterImpl.java
package dds.rmi.example.printer;

import org.opensplice.DDS_RMI.IntHolder;

import printer.JobReport;
import printer.JobStatus;
import printer.PrinterInfos;
import printer.Status;
import DDS_RMI.printer.PrinterInterface;

public class PrinterImpl extends PrinterInterface{

    public static int id = 0;
    private PrinterInfos infos;
    private int activeJobId;

    public PrinterImpl(String name, String location, boolean color){
        infos = new PrinterInfos();
        infos.name = name;
        infos.location = location;
        infos.s = Status.P_IDLE;
        infos.color = color;
        activeJobId = -1;
    }

    synchronized static int getJobId(){
        return id++;
    }

    @Override
    public void suspend() {
        while(getPrinterStatus() != Status.P_IDLE){
            try {
                System.out.println("Waiting for the current job to complete ...");
                Thread.sleep(2000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        System.out.println("Printer " + infos.name + " stopped.");
    }

    @Override
    public void resume() {
        System.out.println("Printer " + infos.name + " activated.");
    }

    @Override
    public JobReport printDoc(String doc_name, boolean color, IntHolder jobId) {
        if(infos.s != Status.P_IDLE){
            return new JobReport(doc_name, infos.name, JobStatus.UNAVAILABLE, (short)0, "The printer is still busy !");
        }
        else
            if(color && !infos.color){
                return new JobReport(doc_name, infos.name, JobStatus.NO_COLOR, (short)0, "The printer is black & white !");
            }

        if(doc_name.length() > 20)
            return new JobReport(doc_name, infos.name, JobStatus.TOO_BIG, (short)0, "The document is too big !");

        setPrinterStatus(Status.P_BUSY);
        jobId.value = getJobId();
        if(jobId.value % 7 == 0){
            setPrinterStatus(Status.P_IDLE);
            return new JobReport(doc_name, infos.name, JobStatus.ERR, (short)0, "Printer error occured !");
        }

        System.out.println("Printing document " + doc_name + " ... ");
        setActiveJob(jobId.value);
        try {
            Thread.sleep(jobId.value * 1000);
        } catch (InterruptedException e) {
            setPrinterStatus(Status.P_IDLE);
            return new JobReport(doc_name, infos.name, JobStatus.ERR, (short)0, "InterruptedException raised !");
        }

        setPrinterStatus (Status.P_IDLE);
        setActiveJob(-1);
        return new JobReport(doc_name, infos.name, JobStatus.PRINTER_OK, (short)doc_name.length(), "Succeeded in printing the document !");
    }

    @Override
    synchronized public int getActiveJob() {
        return activeJobId;
    }

    synchronized private void setActiveJob(int id){
        activeJobId = id;
    }

    synchronized private void setPrinterStatus(Status s){
        infos.s = s;
    }

    synchronized private Status getPrinterStatus(){
        return infos.s;
    }

    @Override
    public PrinterInfos getPrinterInfos() {
        return this.infos;
    }

}
